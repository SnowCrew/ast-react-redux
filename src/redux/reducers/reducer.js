

const initialState = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload,
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const id = action.payload;

      const itemInd = state.itemsInCart.findIndex(item => item.id === id);

      if (itemInd >= 0) {
        const elemInState = state.itemsInCart.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: ++elemInState.counterIdentical
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart.slice(0, itemInd),
            newElem,
            ...state.itemsInCart.slice(itemInd + 1)
          ],
          total: (Number(state.total) + Number(newElem.price)).toFixed(2)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem = {
        title: item.title,
        price: item.price,
        image: item.image,
        id: item.id,
        counterIdentical: 1,
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: (Number(state.total) + Number(newItem.price)).toFixed(2)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload;
      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: state.total - (itemDel.price * itemDel.counterIdentical)
      };

    default:
      return state;
  }
}

export default reducer;