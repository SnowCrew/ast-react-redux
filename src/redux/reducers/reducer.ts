import { IAction } from '../actions/actions';


const initialState:IStore = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0,
  totalQuantity: 0,
  auth: false,
  admin: false
};

export interface IProduct {
  id: number,
  title:string,
  price:number,
  category:string,
  description:string,
  image:string
}

export interface IProductInCart extends IProduct {
counterIdentical:number,
}

export interface IStore {
  products: IProduct[],
  loading: boolean,
  error: boolean,
  itemsInCart: IProductInCart[],
  total: number,
  totalQuantity: number,
  auth: boolean,
  admin: boolean
};

const reducer = (state:IStore = initialState, action:IAction):IStore => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload as IProduct[],
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const [id, quantity] = action.payload as [number, number];
      const itemInd = state.itemsInCart?.findIndex(item => item.id === id);

      if (itemInd >= 0 ) {
        const elemInState = state.itemsInCart?.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: (Number(elemInState?.counterIdentical) + Number(quantity)),
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart!.slice(0, itemInd),
            newElem,
            ...state.itemsInCart!.slice(itemInd! + 1)
          ] as IProductInCart[],
          total: Number((Number(state.total) + (Number(newElem.price) * Number(quantity))).toFixed(2)),
          totalQuantity: Number(state.totalQuantity) + Number(quantity)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem:IProductInCart = {
        title: item!.title,
        price: item!.price,
        image: item!.image,
        id: item!.id,
        category:item!.category,
        description: item!.description,
        counterIdentical: Number(quantity)
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: Number(((state.total) + (Number(newItem.price) * Number(quantity))).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) + Number(quantity)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload as number;

      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      if(itemIndex === -1){
        return state;
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: Number((state.total - (itemDel!.price * itemDel!.counterIdentical)).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) - itemDel!.counterIdentical
      };

    case "CLEAR_CART":
      return {
        ...state,
        itemsInCart: [],
        total: 0,
        totalQuantity: 0
      }

    case "GET_AUTH_REQUEST":
      return {
        ...state,
        loading: true
      };

    case "GET_AUTH":
      return {
        ...state,
        auth: true,
        loading: false
      };

    case "GET_ADMIN_AUTH":
      return {
        ...state,
        auth: true,
        loading: false,
        admin: true
      };

    case "GET_AUTH_REJECT":
      return {
        ...state,
        loading: false
      };

    case "EXIT_AUTH":
      return {
        ...state,
        auth: false,
        admin: false
      };

    default:
      return state;
  };
};

export default reducer;