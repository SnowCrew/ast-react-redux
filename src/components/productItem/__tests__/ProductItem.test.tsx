import { render, screen } from "@testing-library/react";
import { ReactElement } from "react";
import { legacy_createStore as createStore } from "redux";
import ProductItemComponent from "../ProductItem.component";
import reducer, { IProduct } from "../../../redux/reducers/reducer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom/extend-expect";

const renderWithRedux = (
  component: ReactElement,
  { initialState, store = createStore(reducer, initialState) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>, {
      wrapper: MemoryRouter,
    }),
  };
};
const randomProducts1: IProduct = {
  id: 1,
  title: "title1",
  price: 25,
  category: "category1",
  description: "description1",
  image: "image1",
};
const randomProducts2: IProduct = {
  id: 1,
  title: "title2",
  price: 25,
  category: "category2",
  description: "description12",
  image: "image2",
};

describe("ProductItem testing", () => {
  it("ProductItem with auth", () => {
    expect(
      renderWithRedux(<ProductItemComponent product={randomProducts1} />, {
        initialState: {
          products: [randomProducts1, randomProducts2],
          loading: false,
          error: false,
          itemsInCart: [],
          total: 0,
          totalQuantity: 0,
          auth: true,
          admin: false,
        },
      })
    ).toMatchSnapshot();
    expect(screen.getByText("title1")).toBeInTheDocument();
    expect(screen.getByText("Add to cart")).toBeInTheDocument();
    expect(
      screen.queryByText("Чтобы добавить товар в корзину залогинтесь")
    ).not.toBeInTheDocument();
  });

  it("ProductItem without auth", () => {
    renderWithRedux(<ProductItemComponent product={randomProducts1} />, {
      initialState: {
        products: [randomProducts1, randomProducts2],
        loading: false,
        error: false,
        itemsInCart: [],
        total: 0,
        totalQuantity: 0,
        auth: false,
        admin: false,
      },
    });
    expect(
      screen.getByText("Чтобы добавить товар в корзину залогинтесь")
    ).toBeInTheDocument();
    expect(screen.queryByText("Add to cart")).not.toBeInTheDocument();
  });
});
