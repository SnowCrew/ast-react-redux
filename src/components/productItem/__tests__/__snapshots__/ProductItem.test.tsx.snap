// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem testing ProductItem with auth 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="menu__item"
      >
        <a
          class="link"
          href="/1"
        >
          <p
            class="menu__title"
          >
             
            title1
             
          </p>
        </a>
        <img
          alt="title1"
          class="menu__img"
          src="image1"
        />
        <div
          class="menu__category"
        >
          Category: 
          <span>
            category1
          </span>
        </div>
        <div
          class="menu__price"
        >
          Price: 
          <span>
            25
            $
          </span>
        </div>
        <button
          class="menu__btn"
        >
          Add to cart
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="menu__item"
    >
      <a
        class="link"
        href="/1"
      >
        <p
          class="menu__title"
        >
           
          title1
           
        </p>
      </a>
      <img
        alt="title1"
        class="menu__img"
        src="image1"
      />
      <div
        class="menu__category"
      >
        Category: 
        <span>
          category1
        </span>
      </div>
      <div
        class="menu__price"
      >
        Price: 
        <span>
          25
          $
        </span>
      </div>
      <button
        class="menu__btn"
      >
        Add to cart
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
